### DISCUSSION TITLE
Authentication with username and password

### DISCUSSION THREAD

**Discussion by marolo2** [gopcua/opcua#732] [Q&A]
Hello Everyone,
I would like to ask for help with authentication to OPC UA server using username and password.
Below snip of my code how I am creating opc client.

`username := "user"`
`password := "password"`

`opts := []opcua.Option{`
`        opcua.SecurityMode(ua.MessageSecurityModeNone),`
`        opcua.AuthUsername(username, password),`
`        opcua.AutoReconnect(true),`
	`}`
	`ctx := context.Background()`
	`c, _ := opcua.NewClient(opc.tcp://172.17.0.5:4840", opts...)`

but I receive error:
`Failed to connect to OPC UA server: The user identity token is not valid. StatusBadIdentityTokenInvalid (0x80200000)`

When I switch to anonymous mode I can connect and read values, but not at all with password and username. Others clients (written in python or UAExpert) is working fine.

Thank you for help!

**Comment by Dipp3r** [comment]
Try setting the authType to UserName. This should work if the server is accepting username authentication.
For example, 
```
opts := []opcua.Option{
  opcua.SecurityPolicy("None"),
  opcua.SecurityModeString("None"),
  opcua.AuthUsername(username, password),
  opcua.SecurityFromEndpoint(ep, ua.UserTokenTypeUserName),        //I think by default its set to anonymous 
  opcua.AutoReconnect(true),
}
```

"ep" is the endpointDescription returned from the SelectEndpoint method

**Comment by marolo2** [comment]
Hi,
I add `opcua.SelectEndpoint(eps, policy, mode)` in my app and it is working. Below snip:



	endpointUrl := "opc.tcp://172.17.0.5:4840"

	var (
		endpoint = flag.String("endpoint", endpointUrl, "OPC UA Endpoint URL")
	)

	username := "user"
	password := "password"

	eps, err := opcua.GetEndpoints(context.Background(), *endpoint)
	if err != nil {
		log.Fatal(err)
	}

	policy := ua.SecurityPolicyURINone
	mode := ua.MessageSecurityModeNone
	ep := opcua.SelectEndpoint(eps, policy, mode)

	// Create a new OPC UA client with username and password
	opts := []opcua.Option{
		opcua.SecurityFromEndpoint(ep, ua.UserTokenTypeUserName),
		opcua.SessionTimeout(30 * time.Minute),
		opcua.AutoReconnect(true),
		opcua.ReconnectInterval(time.Second * 10),
		opcua.Lifetime(30 * time.Minute),
		opcua.RequestTimeout(3 * time.Second),
		opcua.AuthUsername(username, password),
	}
	ctx := context.Background()

	c, err := opcua.NewClient(*endpoint, opts...)
	if err != nil {
		log.Fatalf("Invalid NewClient: %v", err)
	}

	if err := c.Connect(ctx); err != nil {
		log.Fatalf("Failed to connect to OPC UA server: %v", err)
	}
	defer c.Close(ctx)
